README.md
# NAME

Time::TZOffset - Show timezone offset strings like +0900

# SYNOPSIS

    use Time::TZOffset qw/tzoffset/;

    my @localtime = localtime;
    say tzoffset(@localtime);

# DESCRIPTION

Time::TZOffset provides tzoffset function that determines timezone offset and shows strings
like `+0900`
This module implemented by XS, it's may be faster than other way to show timezone offset.
And also Time::TZOffset is more portable than using `POSIX::strftime` with `%z`.

# FUNCTION

- tzoffset(@localtime)

    Returns a timezone offset string like `+0900`

# BENCHMARK

I did this benchmark on linux.

    use Benchmark qw/:all/;
    use POSIX qw//;
    use Time::Local;
    use Time::TZOffset;
    

    cmpthese(timethese(-1, {
        'posix' => sub {
            POSIX::strftime('%z', @lt);
        },
        'time_local' => sub {
            my $sec = Time::Local::timegm(@lt) - Time::Local::timelocal(@lt);
            sprintf '%+03d%02u', $sec/60/60, $min/60%60;
        },
        'tzoffset' => sub {
            Time::TZOffset::tzoffset(@lt);
        },
    }));
    __END__
    Benchmark: running posix, time_local, tzoffset for at least 1 CPU seconds...
         posix:  1 wallclock secs ( 0.66 usr +  0.39 sys =  1.05 CPU) @ 179442.86/s (n=188415)
    time_local:  1 wallclock secs ( 1.12 usr +  0.16 sys =  1.28 CPU) @ 25846.09/s (n=33083)
      tzoffset:  1 wallclock secs ( 0.75 usr +  0.25 sys =  1.00 CPU) @ 286720.00/s (n=286720)
                   Rate time_local      posix   tzoffset
    time_local  25846/s         --       -86%       -91%
    posix      179443/s       594%         --       -37%
    tzoffset   286720/s      1009%        60%         --

# LICENSE

Copyright (C) Masahiro Nagano.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# POSIX::strftime::GNU LICENSE

This modules uses [POSIX::strftime::GNU](http://search.cpan.org/perldoc?POSIX::strftime::GNU)'s code. [POSIX::strftime::GNU](http://search.cpan.org/perldoc?POSIX::strftime::GNU)'s  license term is following:

Copyright (c) 2012-2014 Piotr Roszatycki <dexter@cpan.org>.

Format specification is based on strftime(3) manual page which is a part of
the Linux man-pages project.

This is free software; you can redistribute it and/or modify it under
the same terms as perl itself.

See [http://dev.perl.org/licenses/artistic.html](http://dev.perl.org/licenses/artistic.html)

# AUTHOR

Masahiro Nagano <kazeburo@gmail.com>


Changes
Revision history for Perl extension Time-TZOffset

{{$NEXT}}

0.04 2015-01-05T03:13:18Z

    - Added offset_as_seconds. (karupanerura)

0.03 2014-05-08T15:09:21Z

    - Correct compilation problems with Visual Studio compiler #1
      (Thank you jacquesg)

0.02 2014-01-23T03:19:34Z

    - use newSVpvf ( thank you typester, gfx)
    - remove time_r 

0.01 2014-01-22T02:50:05Z

    - original version



t/02_timezones.pl
#!/usr/bin/perl

use strict;
use warnings;
use Time::Local;
use Time::TZOffset;

my @t = @ARGV ? localtime timelocal(@ARGV) : localtime;

print Time::TZOffset::tzoffset(@t);



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Time-TZOffset',
    module_name     => 'Time::TZOffset',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw(src)],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


src/time_r.c
/* Reentrant time functions like localtime_r.

   Copyright (C) 2003, 2006-2007, 2010-2012 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, see <http://www.gnu.org/licenses/>.  */

/* Written by Paul Eggert.  */

#include "tzo_config.h"
#include "time_r.h"

static struct tm *
copy_tm_result (struct tm *dest, struct tm const *src)
{
  if (! src)
    return 0;
  *dest = *src;
  return dest;
}


struct tm *
gmtime_r (time_t const * restrict t, struct tm * restrict tp)
{
  return copy_tm_result (tp, gmtime (t));
}

struct tm *
localtime_r (time_t const * restrict t, struct tm * restrict tp)
{
  return copy_tm_result (tp, localtime (t));
}


src/time_r.h
/* declarations for time_r.c

   Copyright (C) 2003, 2006-2007, 2010-2012 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, see <http://www.gnu.org/licenses/>.  */

#include "tzo_config.h"
#include <time.h>

#ifndef HAVE_GMTIME_R
struct tm * gmtime_r (time_t const * restrict t, struct tm * restrict tp);
#endif
#ifndef HAVE_LOCALTIME_R
struct tm * localtime_r (time_t const * restrict t, struct tm * restrict tp);
#endif





lib/Time/TZOffset.pm
package Time::TZOffset;
use 5.008004;
use strict;
use warnings;
use base qw/Exporter/;


our $VERSION = "0.03";
our @EXPORT_OK = qw/tzoffset tzoffset_as_seconds/;

use XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

1;
__END__

=encoding utf-8

=head1 NAME

Time::TZOffset - Show timezone offset strings like +0900

=head1 SYNOPSIS

    use Time::TZOffset qw/tzoffset/;

    my @localtime = localtime;
    say tzoffset(@localtime);

=head1 DESCRIPTION

Time::TZOffset provides tzoffset function that determines timezone offset and shows strings
like C<+0900>
This module implemented by XS, it's may be faster than other way to show timezone offset.
And also Time::TZOffset is more portable than using C<POSIX::strftime> with C<%z>.

=head1 FUNCTION

=over 4

=item tzoffset(@localtime)

Returns a timezone offset string like C<+0900>

=item tzoffset_as_seconds(@localtime)

Returns a timezone offset seconds.

=back

=head1 BENCHMARK

I did this benchmark on linux.

  use Benchmark qw/:all/;
  use POSIX qw//;
  use Time::Local;
  use Time::TZOffset;
  
  cmpthese(timethese(-1, {
      'posix' => sub {
          POSIX::strftime('%z', @lt);
      },
      'time_local' => sub {
          my $sec = Time::Local::timegm(@lt) - Time::Local::timelocal(@lt);
          sprintf '%+03d%02u', $sec/60/60, $min/60%60;
      },
      'tzoffset' => sub {
          Time::TZOffset::tzoffset(@lt);
      },
  }));
  __END__
  Benchmark: running posix, time_local, tzoffset for at least 1 CPU seconds...
       posix:  1 wallclock secs ( 0.66 usr +  0.39 sys =  1.05 CPU) @ 179442.86/s (n=188415)
  time_local:  1 wallclock secs ( 1.12 usr +  0.16 sys =  1.28 CPU) @ 25846.09/s (n=33083)
    tzoffset:  1 wallclock secs ( 0.75 usr +  0.25 sys =  1.00 CPU) @ 286720.00/s (n=286720)
                 Rate time_local      posix   tzoffset
  time_local  25846/s         --       -86%       -91%
  posix      179443/s       594%         --       -37%
  tzoffset   286720/s      1009%        60%         --

=head1 LICENSE

Copyright (C) Masahiro Nagano.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 POSIX::strftime::GNU LICENSE

This modules uses L<POSIX::strftime::GNU>'s code. L<POSIX::strftime::GNU>'s  license term is following:

Copyright (c) 2012-2014 Piotr Roszatycki <dexter@cpan.org>.

Format specification is based on strftime(3) manual page which is a part of
the Linux man-pages project.

This is free software; you can redistribute it and/or modify it under
the same terms as perl itself.

See L<http://dev.perl.org/licenses/artistic.html>

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo@gmail.comE<gt>

=cut

